### DISCUSSION TITLE
Why next state entry action is being executed before  exit action from the previous state?

### DISCUSSION THREAD

**Discussion by mogarick** [statelyai/xstate#2164] [Q&A]
Hi, 
I'm making a PoC using Actors. 
I have this machine: 
``` js
{
  id: "myApp",
  context: {
    currentTask: undefined
  },
  initial: "validatingAuthN",
  states: {
    validatingAuthN: {
      on: {
        SIGNED_IN: { target: "authenticated" },
        SIGNED_OUT: { target: "unauthenticated" }
      }
    },
    unauthenticated: {
      entry: assign({
        currentTask: () => {
          console.log("spawning signInTaskState...");

          return spawn(signInMachine, { sync: true, name: "signInTask" });
        }
      }),
      on: {
        DONE: "authenticated"
      },
      exit: stopCurrentTask //context.currentTask.stop(),
    },
    authenticated: {
      entry: assign({
        currentTask: () => {
          console.log("spawning manageAnimalsTaskFlowState...");
          return spawn(initialTask, {
            sync: true,
            name: initialTask.id
          });
        }
      }),

      on: {
        SIGNED_OUT: "unauthenticated",
        AUTHN_VALIDATION_REQUESTED: "validatingAuthN"
      },
      exit: stopCurrentTask
    }
  }
};
```

I'm having a weird problem that I don't know if its due I'm new to xstate or if there is a bug. Here is the scenario:
Starting from `validatingAuthN`
1. Trigger `SIGNED_OUT`. it goes to `unauthenticated` state, spawning signInMachine in the `entry` action.
2. Trigger `DONE` while being on `unauthenticated` state. I'd expect `unauthenticated ` `exit` action to be executed (where I explicitly stop the currentTask machine) and then go to `authenticated` state. But what happens is the `authenticated ` entry action is executed first, spawning the initialTaskMachine and then the `exit` action. The problem with this is that at that point the currentTask has changed to initialTaskMachine and this machine becomes the one stoped when the one I expected to be stoped was signInTask.

It's important to note that the problem also occurs if I first trigger SIGN_IN and then trigger SIGNED_OUT while on the signInTask machine. The entry action from the next state is executed before de exit action from the previous state.

Is there something I'm doing wrong that is causing this weird behavior?

Here is the codeSandbox with the full scenario.

https://codesandbox.io/s/mogarick-spawn-svsib?file=/src/index.ts:5160-5171

Thank you in advance for your help.

**Comment by ChrisShank** [comment]
Not sure why the order execution is different than I would expected here... but it sort of feels like you should `invoke` these machines instead of spawn them, since the machines should only exist in the while in a certain state! Would this work in your case?

```ts
const machine = createMachine({
  id: "anivettyApp",
  initial: "validatingAuthN",
  states: {
    validatingAuthN: {
      on: {
        SIGNED_IN: { target: "authenticated" },
        SIGNED_OUT: { target: "unauthenticated" }
      }
    },
    unauthenticated: {
      invoke: {
        id: 'signIn',
        src: signInMachine,
        onDone: 'authenticated'
      },
    },
    authenticated: {
      invoke: {
        id: 'task', 
        src: initialTask,
      },
      on: {
        SIGNED_OUT: "unauthenticated",
        AUTHN_VALIDATION_REQUESTED: "validatingAuthN"
      },
    }
  }
});
```

**Reply by mogarick** [reply]
Invoke was my first experiment for this. The problem is I don't have a way to track the state changes for such invoked machines. I can only track parent state changes and that's not what I need. I've been searching for ways to track states changes in the child machines without success. Can you point me out to a link here I can see how can be done or isn't it possible?

**Reply by ChrisShank** [reply]
Can you clarify what you mean by "track the state changes for such invoked machines"?

**Reply by mogarick** [reply]
I mean getting to now the invoked machine states as it transitions between them. `onTransition` only tracks state changes for the parent machine. But I have no way to know about the invoked machine state transitions. I'm not necessarily interested in knowing them from the parent but from the invoked machine itself. Maybe something like:

```
let myChileMachine = Machine(config).onTransition(state=>console.log("Transitioned to state:",state.value))
```

This is no possible AFAIK. `onTransition` is only available as part of Interpretrer and Interpreter is only used for interpreting the Parent Machine. The child Machines get processed in that parent interpretation but there is no way to listen for their internal state transitions.

I need to know about the child machine transitions but in a way independent of UI son I can't depend on `useMachine` nor `useService`, etc.

**Reply by ChrisShank** [reply]
I see, i think you can subscribe/unsubscribe to invoked or spawned actors! You just have to make sure to unsubscribe to the child actor if it doesn't exist anymore.
```ts
let childSubscription = null;

const service = interpret(...).onTransition(state => {
  // const child = state.context.actorRef
  const child = state.children.invokedChild

  if (childSubscription) {
    childSubscription.unsubscribe()
    childSubscription = null;
  }

  if (child) {
    childSubscription = child.subscribe(state => {
      // listen to chili actor
    })
  }
})
```

**Comment by Andarist** [comment]
As to the execution order - this is something that we'll be fixing in v5. Your exit actions are "custom" right now (not based on built-in ones) and they are executed after the internal `transition` call, whereas built-in actions are "resolved" within that `transition` call.

You also shouldn't call `.stop()` on spawned actors directly, but rather use a `stop` action for that (otherwise you are risking memory leaks). You can learn more about this here: https://github.com/davidkpiano/xstate/pull/1434

Here you have the corrected sandbox: https://codesandbox.io/s/mogarick-spawn-forked-6nzie?file=/src/index.ts

**Reply by mogarick** [reply]
Thank you for the `.stop()` tip. 

Regarding the exit actions, does it mean that this "out-of-expected-order" has no solution at this moment in version 4.x? Is there an alternative to make this work as I'm expecting?

Finally, the fact that I can still see the machine on the visualizer even though it's been stopped is something that will also be fixed or is there an alternate way to make sure it doesn't show up once it's been stoped?

<img width="345" alt="image" src="https://user-images.githubusercontent.com/4414145/116883601-67ad3180-abeb-11eb-8666-e32c69badcd7.png">

Thank you again for your attention! :)

**Reply by Andarist** [reply]
> Regarding the exit actions, does it mean that this "out-of-expected-order" has no solution at this moment in version 4.x? Is there an alternative to make this work as I'm expecting?

This is a little bit tricky because in v4 assigns are prioritized. You can kinda work around this by using `pure`  like here:
https://codesandbox.io/s/mogarick-spawn-forked-lqc21?file=/src/index.ts

Note that this still kinda executes those out of order **but** you would get access to the captured values from the moment when `pure` has been resolved.

> Finally, the fact that I can still see the machine on the visualizer even though it's been stopped is something that will also be fixed or is there an alternate way to make sure it doesn't show up once it's been stoped?

I can't reproduce this on the codesandbox I've created for this post.

**Reply by mogarick** [reply]
> Note that this still kinda executes those out of order but you would get access to the captured values from the moment when pure has been resolved.

Indeed.  I've tried it out. Thank you for the tip.

> I can't reproduce this on the codesandbox I've created for this post.

You can reproduce it only if you transition the states directly in the inspect state chart. If you do it programmatically it can't be reproduced.

<img width="1664" alt="image" src="https://user-images.githubusercontent.com/4414145/117191125-d7671c00-ada5-11eb-8906-1b1fff11955d.png">
Programatic transition (`s.send({ type: "NEXT" }); s.send({ type: "NEXT" });`)

<img width="1673" alt="image" src="https://user-images.githubusercontent.com/4414145/117204328-cb835600-adb5-11eb-89be-d66b882e5f40.png">
Manual transition (clicking directly on the state chart's events)

**Reply by Andarist** [reply]
> You can reproduce it only if you transition the states directly in the inspect state chart.

Could you report this as an issue?

**Reply by mogarick** [reply]
Sorry for the delay. I've just added the scenario for this problem to issue #2135 reported previously by me and also related to the inspector weird behavior for nested machines.

Thank you for your help. 💯
